pkg/metrics/store.go

```diff
@@ -184,10 +185,31 @@ func (s *metrics) Histogram(key string) gometrics.Histogram {
 		return gometrics.NilHistogram{}
 	}
 
-	construct := func() gometrics.Histogram {
-		// TODO: notice the histogram only keeps 100 values as we set
-		return s.registry.GetOrRegister(key, func() gometrics.Histogram { return gometrics.NewHistogram(gometrics.NewUniformSample(100)) }).(gometrics.Histogram)
+	var construct func() gometrics.Histogram
+	switch sampleType {
+	case SampleUniform:
+		construct = func() gometrics.Histogram {
+			return s.registry.GetOrRegister(key, func() gometrics.Histogram {
+				return gometrics.NewHistogram(gometrics.NewUniformSample(sampleSize))
+			}).(gometrics.Histogram)
+		}
+	case SampleExpDecay:
+		construct = func() gometrics.Histogram {
+			return s.registry.GetOrRegister(key, func() gometrics.Histogram {
+				return gometrics.NewHistogram(gometrics.NewExpDecaySample(sampleSize, expDecayAlpha))
+			}).(gometrics.Histogram)
+		}
+	default:
+		if log.DefaultLogger.GetLogLevel() >= log.WARN {
+			log.DefaultLogger.Warnf("[metrics] Unknown sample type %s, will use UNIFORM as default", sampleType)
+		}
```

Comment:
It would be better to do the fallback in SetSampleType, instead of when we need to use it?

Reply:
Hello, @spacewander . 

I agree with you, but I don't like doing too much in the setters. And I realized that it shouldn't be limited to only these two types. In the case of package references (such as layotto), users should be able to expand more sampling logic. 

So I did some refactoring to support user registration of custom sampling types 😺

Reply:
If the user configures an incorrect value in the configuration (probably via SetSampleType), we should reject it at the beginning, instead of throwing a warning when creating each upstream/downstream stat.

Reply:
I understand what you mean, I will check if the type exists in the map in the setter